### PR TITLE
#111 sofort iban handling

### PR DESCRIPTION
**Pull request by mht-dotsource**
#111 #109 #94 #93 #55

### CODE REVIEW

**Comment by andrii-kovalenko-ct** [comment]
@mht-dotsource 
when could you merge the branch? I need changes from `gradle.build` in master (updated version of jvm sdk)

**Comment by mht-dotsource** [comment]
@andrii-kovalenko-ct 
I issued the PR someone else should merge it. Can you contact @floriansattler when it will be reviewed?

File: functionaltests/src/test/java/specs/response/ResponsesFixture.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+                        .build()));
+
+        return payment.getId();
+    }
+
+    public Map<String, String> handlePayment(final String paymentName,
+                                             final String requestType) throws ExecutionException, IOException {
+        final HttpResponse response = requestToHandlePaymentByLegibleName(paymentName);
+        final Payment payment = fetchPaymentByLegibleName(paymentName);
+        final String transactionId = getIdOfFirstTransaction(payment);
```
Comment (andrii-kovalenko-ct) [inline functionaltests/src/test/java/specs/response/ResponsesFixture.java]: No null pointer check. May be some additional `assertNotNull` should be added.

File: service/src/main/java/com/commercetools/pspadapter/payone/mapping/PayoneRequestFactory.java [hunk 2]
```diff
@@ -33,4 +36,31 @@ public CaptureRequest createCaptureRequest(final PaymentWithCartLike paymentWith
...
+            MappingUtil.mapBillingAddressToRequest(request, ctCartLike.getBillingAddress());
+        } catch (final IllegalArgumentException ex) {
+            logger.warn("Could not fully map payment with ID " + paymentWithCartLike.getPayment().getId(), ex.getMessage());
+        }
+
+        try {
+            MappingUtil.mapShippingAddressToRequest(request, ctCartLike.getShippingAddress());
+        } catch (final IllegalArgumentException ex) {
+            logger.warn("Could not fully map payment with ID " + paymentWithCartLike.getPayment().getId(), ex.getMessage());
+        }
```
Comment (andrii-kovalenko-ct) [inline service/src/main/java/com/commercetools/pspadapter/payone/mapping/PayoneRequestFactory.java]: 1. Looks like may be `static`

2. Why only `mapBillingAddressToRequest` logged as an _error_, while all other as _debug_?

File: service/src/main/java/com/commercetools/pspadapter/payone/mapping/PostFinanceBanktransferRequestFactory.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+    public PostFinanceBanktransferRequestFactory(PayoneConfig payoneConfig) {
+        super(payoneConfig);
+    }
+
+    @Override
+    public BankTransferAuthorizationRequest createAuthorizationRequest(PaymentWithCartLike paymentWithCartLike) {
+        final BankTransferAuthorizationRequest request = super.createAuthorizationRequest(paymentWithCartLike);
+        //Despite declared as optional in PayOne Server API documentation. this is required!
+        request.setBankcountry("CH");
```
Comment (andrii-kovalenko-ct) [inline service/src/main/java/com/commercetools/pspadapter/payone/mapping/PostFinanceBanktransferRequestFactory.java]: Better to have some constant/enum for "CH" (and potentially other values)

File: service/src/main/java/com/commercetools/pspadapter/payone/notification/NotificationProcessorBase.java [hunk 4]
```diff
@@ -72,8 +74,14 @@ public final void processTransactionStatusNotification(final Notification notifi
...
      * or apply changes to a corresponding transaction in the payment
      */
-    protected abstract ImmutableList<UpdateAction<Payment>> createPaymentUpdates(final Payment payment,
-                                                                                 final Notification notification);
+    protected ImmutableList<UpdateAction<Payment>> createPaymentUpdates(final Payment payment,
+                                                                                 final Notification notification) {
+        final ImmutableList.Builder<UpdateAction<Payment>> listBuilder = ImmutableList.builder();
+        listBuilder.add(createNotificationAddAction(notification));
+        listBuilder.add(setStatusInterfaceCode(notification));
+        listBuilder.add(setStatusInterfaceText(notification));
```
Comment (andrii-kovalenko-ct) [inline service/src/main/java/com/commercetools/pspadapter/payone/notification/NotificationProcessorBase.java]: Is [`PaymentStatus#state`](http://dev.commercetools.com/http-api-projects-payments.html#paymentstatus) (not transaction state) set somewhere? Looks like it is missed from the payment transaction response.

File: service/src/main/java/com/commercetools/pspadapter/payone/transaction/TransactionBaseExecutor.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+/**
+ * @author mht@dotsource.de
+ * Common base class responsible for default paymentupdateactions
+ */
+public abstract class TransactionBaseExecutor extends IdempotentTransactionExecutor{
+
+    private final static String STATUS = "status";
+    private final static String ERROR_CODE = "errorcode";
+    private final static String CUSTOMER_MESSAGE = "customermessage";
```
Comment (andrii-kovalenko-ct) [inline service/src/main/java/com/commercetools/pspadapter/payone/transaction/TransactionBaseExecutor.java]: Created, but never used. Or `PaymentStatus` should be expanded, or `CustomField` should be created.

File: service/src/main/java/com/commercetools/pspadapter/payone/transaction/TransactionBaseExecutor.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+     * @param response contains all key that creates the
+     * @return the UpdateAction that
+     */
+    protected SetStatusInterfaceCode setStatusInterfaceCode(final Map<String, String> response){
+        final String status = response.get(STATUS);
+        final StringBuilder stringBuilder = new StringBuilder(status);
+        if(ERROR.equals(status)) {
+            //Payone errorcode is required for error case
+            stringBuilder.append(" ");
+            stringBuilder.append(response.get(ERROR_CODE));
```
Comment (mht-dotsource) [inline service/src/main/java/com/commercetools/pspadapter/payone/transaction/TransactionBaseExecutor.java]: @andrii-kovalenko-ct you make a good point that processing it later would be easier when its saved seperately. Concatinating was described in the Link in Issue #55

File: service/src/main/java/com/commercetools/pspadapter/payone/transaction/TransactionBaseExecutor.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    protected SetStatusInterfaceText setStatusInterfaceText(final Map<String, String> response){
+        final String status = response.get(STATUS);
+        final StringBuilder stringBuilder = new StringBuilder("");
+        if(ERROR.equals(status)) {
+            stringBuilder.append(response.get(ERROR_MESSAGE));
+        }
+        else {
+            stringBuilder.append(status);
+        }
+        return SetStatusInterfaceText.of(stringBuilder.toString());
```
Comment (mht-dotsource) [inline service/src/main/java/com/commercetools/pspadapter/payone/transaction/TransactionBaseExecutor.java]: yes indeed could be simplified

File: functionaltests/src/test/java/specs/response/ResponsesFixture.java (hunk 1 above)
Comment (mht-dotsource) [inline functionaltests/src/test/java/specs/response/ResponsesFixture.java]: Yes would make the test more robust.

File: service/src/main/java/com/commercetools/pspadapter/payone/mapping/PayoneRequestFactory.java (hunk 2 above)
Comment (mht-dotsource) [inline service/src/main/java/com/commercetools/pspadapter/payone/mapping/PayoneRequestFactory.java]: Indeed could be improved even more. Just refactored the code block to not be repeated in every Requestfactory

File: service/src/main/java/com/commercetools/pspadapter/payone/mapping/PostFinanceBanktransferRequestFactory.java (hunk 3 above)
Comment (mht-dotsource) [inline service/src/main/java/com/commercetools/pspadapter/payone/mapping/PostFinanceBanktransferRequestFactory.java]: yes can be extrected to constant

**Comment by mht-dotsource** [comment]
@andrii-kovalenko-ct 

The usage of the fields for are specified in https://github.com/commercetools/commercetools-payone-integration/blob/master/docs/Field-Mapping.md#commercetools-payment-resource-mapping

1. Therefore the String was concatinated with the message.
2. PaymentStatus#state is supposed to be left empty because its could be used project specific.
3. Customer Message is indeed not mapped (can only appear in ERROR response) and should be mapped to Interface.Text

**Pull request merged**